lib/helpers/versions.js

```diff
@@ -2,15 +2,84 @@ const logger = require('winston');
 const semver = require('semver');
 const stable = require('semver-stable');
 const _ = require('lodash');
+const semverUtils = require('semver-utils');
 
 module.exports = {
+  determineRangeUpgrades,
   determineUpgrades,
   isRange,
   isValidVersion,
   isFuture,
   isPastLatest,
+  computeRangedVersion,
 };
 
+function determineRangeUpgrades(dep, currentVersion, config) {
+  if (!isValidVersion(currentVersion)) {
+    logger.verbose(`${dep.name} currentVersion is invalid`);
+    return [];
+  }
+  const versions = dep.versions;
+  if (!versions || Object.keys(versions).length === 0) {
+    logger.verbose(`${dep.name} - no versions`);
+    return [];
+  }
+  const versionList = Object.keys(versions);
+  const workingVersion = semver.maxSatisfying(versionList, currentVersion);
+
+  return _.chain(versionList)
+
+    // Filter out older versions as we can't upgrade to those
+    .filter(version => semver.gt(version, workingVersion))
+
+    // Ignore unstable versions, unless the current version is unstable
+    .reject(version => config.ignoreUnstable &&
+            stable.is(workingVersion) && !stable.is(version))
+
+    // Ignore future versions, unless the current version is marked as future
+    .reject(version => config.ignoreFuture &&
+            !isFuture(versions[workingVersion]) && isFuture(versions[version]))
+
+    // Ignore versions newer than "latest", unless current version is newer than the "latest"
+    .reject(version => config.respectLatest &&
+            isPastLatest(dep, version) && !isPastLatest(dep, workingVersion))
+
+    // Process all remaining versions
+    .reduce((upgrades, newVersion) => {
+      const parsedVersion = semver.parse(newVersion);
+      const newUpgrade = {};
+
+      // Group updates for the current dependency by the major version.
+      if (!upgrades[parsedVersion.major] ||
+          semver.gt(newVersion, upgrades[parsedVersion.major].newVersion)) {
```

Comment:
The reason I keep a separate `newVersion` and `newVersionRange` is because of this check. `semver.gt` only works on a pinned version.

Please feel free to push a commit to replace the use of `newVersion` and `newVersionRange` with any naming scheme you feel is appropriate.